### PR TITLE
Interface aliases

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add the ability to create interface aliases

### CODE REVIEW

**Comment by phlptp** [comment]
This will not include the API for C shared library, that will come in a subsequent PR.

**Comment by phlptp** [comment]
I think I got most of the tests in.  Going to add a documentation page, and maybe some tests for the file interpreters

**Comment by phlptp** [comment]
Two things I will need to do **NOT** in this PR.  
1). more comprehensive error handling and tests with aliases
2).  Add the C API for aliases.

File: src/helics/core/UnknownHandleManager.hpp [hunk 1]
```diff
@@ -108,6 +108,14 @@ class UnknownHandleManager {
     */
     void processNonOptionalUnknowns(
         const std::function<void(const std::string& name, char type, GlobalHandle)>& cfunc) const;
+
+    /** run a callback for each Unknown
+    @param cfunc a callback function with the signature of the name of the required interface a
+    character with the type 'p' for publication, 'i' for input, 'f' for filter, 'e' for endpoint and
+    the global handle.
```
Comment (nightlark) [inline src/helics/core/UnknownHandleManager.hpp:115]:
```suggestion
    @param cfunc a callback function with the signature of the name of the required interface, a
    character with the type ('p' for publication, 'i' for input, 'f' for filter, 'e' for endpoint), and
    the global handle.
```

File: tests/helics/core/HandleManagerTests.cpp [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+    h1.addAlias("p1", "pub");
+    h1.addAlias("pub", "publisher");
+
+    const auto* p1 = h1.getPublication("publisher");
+    ASSERT_NE(p1, nullptr);
+    EXPECT_EQ(p1->key, "p1");
+
+    const auto* p2 = h1.getPublication("pub");
+    ASSERT_NE(p2, nullptr);
+    EXPECT_EQ(p1->key, "p1");
```
Comment (nightlark) [inline tests/helics/core/HandleManagerTests.cpp:528]:
```suggestion
    EXPECT_EQ(p2->key, "p1");
```

Unless getting a publication alters p1, this is a duplicate of the check on line 524.

File: src/helics/core/HandleManager.cpp [hunk 3]
```diff
@@ -364,27 +366,233 @@ LocalFederateId HandleManager::getLocalFedID(InterfaceHandle handle) const
...
+    auto fnd = endpoints.find(interfaceName);
+    if (fnd != endpoints.end()) {
+        auto res = endpoints.try_emplace(alias, fnd->second.baseValue());
+        if (!res.second && res.first->second != fnd->second) {
+            throw std::runtime_error("endpoint name already exists");
+        }
+    } else {
+        fnd = publications.find(alias);
+        if (fnd != publications.end()) {
+            publications.emplace(interfaceName, fnd->second.baseValue());
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/HandleManager.cpp:426]: yes

File: src/helics/application_api/Federate.cpp [hunk 4]
```diff
@@ -1131,6 +1131,20 @@ void Federate::registerConnectorInterfacesJson(const std::string& jsonString)
...
+    if (doc.isMember("aliases")) {
+        if (doc["aliases"].isArray()) {
+            for (auto& val : doc["aliases"]) {
+                addAlias(val[0].asString(), val[1].asString());
+            }
+        } else {
+            auto members = doc["aliases"].getMemberNames();
+            for (auto& val : members) {
+                addAlias(val, doc["aliases"][val].asString());
+            }
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/application_api/Federate.cpp:1144]: I will add this when I add the C API for aliases

File: src/helics/core/ActionMessageDefintions.hpp [hunk 5]
```diff
@@ -122,6 +122,7 @@ across different compilers*/
             cmd_info_basis + 707,  //!< command to connect a publication with an endpoint
         cmd_filter_link = cmd_info_basis + 709,  //!< command to add a target to a filter
         cmd_endpoint_link = cmd_info_basis + 711,  //!< command to link endpoints
+        cmd_add_alias = 715,  //!< command to add an alias for an interface
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/ActionMessageDefintions.hpp]: I added the cmd_info_basis.  I am pretty sure this mattered in HELICS 2 when that was a marker for an extra string.  That changed in HELICS3 but I didn't change the command code numbers so it doesn't matter anymore.

File: src/helics/core/CoreBroker.hpp [hunk 6]
```diff
@@ -330,11 +331,11 @@ class CoreBroker: public Broker, public BrokerBase {
...
     /** find any existing publishers for a subscription*/
-    void FindandNotifyInputTargets(BasicHandleInfo& handleInfo);
-    void FindandNotifyPublicationTargets(BasicHandleInfo& handleInfo);
+    void FindAndNotifyInputTargets(BasicHandleInfo& handleInfo, const std::string& key);
+    void FindAndNotifyPublicationTargets(BasicHandleInfo& handleInfo, const std::string& key);
 
-    void FindandNotifyFilterTargets(BasicHandleInfo& handleInfo);
-    void FindandNotifyEndpointTargets(BasicHandleInfo& handleInfo);
+    void FindAndNotifyFilterTargets(BasicHandleInfo& handleInfo, const std::string& key);
+    void FindAndNotifyEndpointTargets(BasicHandleInfo& handleInfo, const std::string& key);
```
Comment (nightlark) [inline src/helics/core/CoreBroker.hpp:338]: If we want to match the naming convention used by other functions, the first letter of these function names should be lowercase.

**Review by nightlark — APPROVED** [review]

**Pull request merged**